### PR TITLE
MPI related packages: use multiple outputs

### PR DESCRIPTION
**Pull request by markuskowa**
## Description of changes
Separate  runtime file from documentation/man pages, and build time files (where possible).
Note that for the MPI packages the `dev` output cannot easily be separated as this leads to cyclic dependencies between `out` and `dev`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by markuskowa** [comment]
@GrahamcOfBorg test slurm

File: pkgs/development/libraries/openmpi/default.nix [hunk 1]
```diff
@@ -66,13 +68,13 @@ in stdenv.mkDerivation rec {
...
     ++ lib.optionals cudaSupport [ "--with-cuda=${cudatoolkit_joined}" "--enable-dlopen" ]
-    ++ lib.optionals fabricSupport [ "--with-psm2=${libpsm2}" "--with-libfabric=${libfabric}" ]
+    ++ lib.optionals fabricSupport [ "--with-psm2=${lib.getDev libpsm2}" "--with-libfabric=${lib.getDev libfabric}" ]
     ;
 
   enableParallelBuilding = true;
 
   postInstall = ''
-    rm -f $out/lib/*.la
+    find $out/lib/ -name "*.la" -exec rm -f \{} \;
```
Comment (bjornfor) [inline pkgs/development/libraries/openmpi/default.nix:77]: Unrelated change. Please put it in a separate commit and explain the what/why of that change.

(I checked and found that there is only one *.la file in the output: `result/lib/openmpi/libompi_dbg_msgq.la`. So that means the current code is a NOP. I'm not sure why we're removing *.la files though.)

File: pkgs/development/libraries/pmix/default.nix [hunk 2]
```diff
@@ -23,15 +23,19 @@ stdenv.mkDerivation rec {
...
-    "--with-hwloc=${hwloc.dev}"
+    "--with-hwloc=${lib.getDev hwloc}"
   ];
 
   preConfigure = ''
     ./autogen.pl
   '';
 
+  postInstall = ''
+    find $out/lib/ -name "*.la" -exec rm -f \{} \;
```
Comment (bjornfor) [inline pkgs/development/libraries/pmix/default.nix:36]: Can you say why removing .la files is a good thing in the commit message?

File: pkgs/development/libraries/openmpi/default.nix (hunk 1 above)
Comment (markuskowa) [reply pkgs/development/libraries/openmpi/default.nix:77]: It actually has la files in `lib/` and the one in `lib/openmpi` (the old version just removed them):
```
result/lib/libmpi.la
result/lib/libmpi_usempif08.la
result/lib/libompitrace.la
result/lib/openmpi/libompi_dbg_msgq.la
result/lib/libmpi_usempi_ignore_tkr.la
result/lib/libmpi_mpifh.la
result/lib/liboshmem.la
result/lib/libopen-pal.la
result/lib/libmpi_cxx.la
result/lib/libopen-rte.la
```

A far as I am aware, these libtool files are not needed for anything on Linux (or on Darwin). The idea here is to simply to reduce the output size/number of files in the output.

Comment (bjornfor) [reply pkgs/development/libraries/openmpi/default.nix:77]: Hehe, silly me, the rm code worked :-)

**Review by bzizou — APPROVED** [review]
LGTM
